# Analysis of [Major League Baseball](https://en.wikipedia.org/wiki/Major_League_Baseball) data

In this project we will be exploring data about [Major league baseball](https://en.wikipedia.org/wiki/Major_League_Baseball) games compiled by [Retrosheet](http://www.retrosheet.org/). Retrosheet compiles detailed statistics on baseball games from the 1800s through to today. The main file we will be working from **`game_log.csv`**


Lets start by familiarizing ourselves with columns in **`game.csv`** . We will also remove any older version of mlb.db 

In [1]:
!cat game_fields.txt

# from IPython.core.interactiveshell import InteractiveShell
# InteractiveShell.ast_node_interactivity = "all"

'cat' is not recognized as an internal or external command,
operable program or batch file.


In [2]:
!rm mlb.db

'rm' is not recognized as an internal or external command,
operable program or batch file.


In [3]:
import pandas as pd

pd.set_option('max_columns', 180)
pd.set_option('max_rows', 200000)
pd.set_option('max_colwidth', 5000)

## Exploratory data analsys

### Start reading data files.

We will read each file separately and dump some rows and shape to understand the kind of contained in the dataframe. We want to get a general sense of what data is contained in these tables and how they relate to each other

In [4]:
game = pd.read_csv("game_log.csv")
game.shape



C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (12,13,14,15,19,20,81,82,83,84,85,86,87,88,93,94,95,96,97,98,99,100,105,106,108,109,111,112,114,115,117,118,120,121,123,124,126,127,129,130,132,133,135,136,138,139,141,142,144,145,147,148,150,151,153,154,156,157,160) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(171907, 161)

In [5]:
game.head(3)

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1      NaN              1    FW1   
1  18710505               0         Fri    BS1      NaN              1    WS3   
2  18710506               0         Sat    CL1      NaN              2    RC1   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0      NaN              1        0        2         54.0         D        NaN   
1      NaN              1       20       18         54.0         D        NaN   
2      NaN              1       12        4         54.0         D        NaN   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0      NaN     NaN   FOR01       200.0           120.0    000000000   
1      NaN     NaN   WAS01      5000.0           145.0    107000435   
2      NaN     NaN   RCK01      1000.0           140.0    610020003   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   

   v_intentional_walks  v_strikeouts  v_stolen_bases  v_caught_stealing  \
0                  NaN           6.0             1.0                NaN   
1                  NaN           5.0             3.0                NaN   
2                  NaN           1.0             0.0                NaN   

   v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
0                    -1.0                           NaN             4.0   
1                    -1.0                           NaN            12.0   
2                    -1.0                           NaN            10.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   

   v_double_plays  v_triple_plays  h_at_bats  h_hits  h_doubles  h_triples  \
0             0.0             0.0       31.0     4.0        1.0        0.0   
1             2.0             0.0       49.0    14.0        2.0        0.0   
2             0.0             0.0       36.0     7.0        2.0        1.0   

   h_homeruns  h_rbi  h_sacrifice_hits  h_sacrifice_flies  h_hit_by_pitch  \
0         0.0    2.0               0.0                0.0             0.0   
1         0.0   11.0               0.0                0.0             0.0   
2         0.0    2.0               0.0                0.0             0.0   

   h_walks  h_intentional_walks  h_strikeouts  h_stolen_bases  \
0      1.0                  NaN           0.0             0.0   
1     10.0                  NaN           2.0             1.0   
2      0.0                  NaN           3.0             5.0   

   h_caught_stealing  h_grounded_into_double  h_first_catcher_interference  \
0                NaN                    -1.0                           NaN   
1                NaN                    -1.0                           NaN   
2                NaN                    -1.0                           NaN   

   h_left_on_base  h_pitchers_used  h_individual_earned_runs  \
0             3.0              1.0        

In [6]:
game.tail(3)

date  number_of_game day_of_week v_name v_league  v_game_number  \
171904  20161002               0         Sun    LAN       NL            162   
171905  20161002               0         Sun    PIT       NL            162   
171906  20161002               0         Sun    MIA       NL            161   

       h_name h_league  h_game_number  v_score  h_score  length_outs  \
171904    SFN       NL            162        1        7         51.0   
171905    SLN       NL            162        4       10         51.0   
171906    WAS       NL            162        7       10         51.0   

       day_night completion forefeit protest park_id  attendance  \
171904         D        NaN      NaN     NaN   SFO03     41445.0   
171905         D        NaN      NaN     NaN   STL10     44615.0   
171906         D        NaN      NaN     NaN   WAS11     28730.0   

        length_minutes v_line_score h_line_score  v_at_bats  v_hits  \
171904           184.0    000100000    23000002x       30.0     4.0   
171905           192.0    000020200    00100360x       35.0     9.0   
171906           216.0    000230020    03023002x       38.0    14.0   

        v_doubles  v_triples  v_homeruns  v_rbi  v_sacrifice_hits  \
171904        0.0        0.0         0.0    1.0               0.0   
171905        0.0        0.0         1.0    4.0               0.0   
171906        1.0        1.0         2.0    7.0               1.0   

        v_sacrifice_flies  v_hit_by_pitch  v_walks  v_intentional_walks  \
171904                0.0             0.0      2.0                  0.0   
171905                0.0             0.0      4.0                  0.0   
171906                0.0             0.0      3.0                  2.0   

        v_strikeouts  v_stolen_bases  v_caught_stealing  \
171904           7.0             0.0                0.0   
171905          11.0             0.0                1.0   
171906          10.0             1.0                1.0   

        v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
171904                     1.0                           0.0             4.0   
171905                     0.0                           0.0             8.0   
171906                     1.0                           0.0             8.0   

        v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
171904              7.0                       7.0                 7.0   
171905              6.0                       8.0                 8.0   
171906              7.0                      10.0                10.0   

        v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  \
171904             0.0      0.0       24.0        5.0       1.0   
171905             0.0      0.0       24.0        2.0       2.0   
171906             1.0      0.0       24.0       11.0       0.0   

        v_passed_balls  v_double_plays  v_triple_plays  h_at_bats  h_hits  \
171904             0.0             0.0             0.0       39.0    16.0   
171905             0.0             0.0             0.0       36.0    12.0   
171906             0.0             1.0             0.0       30.0    10.0   

        h_doubles  h_triples  h_homeruns  h_rbi  h_sacrifice_hits  \
171904        3.0        1.0         0.0    7.0               0.0   
171905        2.0        0.0         1.0   10.0               0.0   
171906        2.0        0.0         1.0   10.0               1.0   

        h_sacrifice_flies  h_hit_by_pitch  h_walks  h_intentional_walks  \
171904                0.0             0.0      4.0                  1.0   
171905                2.0             0.0      4.0                  0.0   
171906                1.0             1.0      8.0                  0.0   

        h_strikeouts  h_stolen_bases  h_caught_stealing  \
171904          11.0             2.0                1.0   
171905           5.0             0.0                0.0   
171906           3.0             2.0                0.0   

        h_grounded_into_do

It looks like the game log has a record of over 170,000 games. It looks like these games are chronologically ordered and occur between 1871 and 2016.

For each game we have:

   - general information on the game
   - team level stats for each team
   - a list of players from each team, numbered, with their defensive positions
   - the umpires that officiated the game
   - some 'awards', like winning and losing pitcher

We have a `game_fields.txt` file that tell us that the player number corresponds with the order in which they batted.

It's worth noting that there is no natural primary key column for this table.

In [7]:
park = pd.read_csv("park_codes.csv")
park.shape



(252, 9)

In [8]:
park.head(10)

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   
5   ATL01  Atlanta-Fulton County Stadium   
6   ATL02                   Turner Field   
7   ATL03                  Suntrust Park   
8   BAL01         Madison Avenue Grounds   
9   BAL02                 Newington Park   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   
5                                       NaN    Atlanta    GA  04/12/1966   
6                                       NaN    Atlanta    GA  04/04/1997   
7                                       NaN    Atlanta    GA  04/14/2017   
8                                       NaN  Baltimore    MD  07/08/1871   
9                                       NaN  Baltimore    MD  04/22/1872   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN  
5  09/23/1996     NL                                             NaN  
6  10/02/2016     NL                                             NaN  
7         NaN     NL                                             NaN  
8  07/08/1871    NaN                                             WS3  
9  09/30/1882    NaN                BL1:1872-74; BL4:1873; BL2: 1882

In [9]:
park.tail(10)

park_id                                 name  aka           city state  \
242   WAT01               Troy Ball Club Grounds  NaN     Watervliet    NY   
243   WAV01                  Waverly Fairgrounds  NaN        Waverly    NJ   
244   WEE01                      Monitor Grounds  NaN      Weehawken    NJ   
245   WHE01                       Island Grounds  NaN       Wheeling    WV   
246   WIL01                    Union Street Park  NaN     Wilmington    DE   
247   WIL02        BB&T Ballpark at Bowman Field  NaN    Wiliamsport    PA   
248   WNY01     West New York Field Club Grounds  NaN  West New York    NJ   
249   WOR01   Agricultural County Fair Grounds I  NaN      Worcester    MA   
250   WOR02  Agricultural County Fair Grounds II  NaN      Worcester    MA   
251   WOR03       Worcester Driving Park Grounds  NaN      Worcester    MA   

          start         end league  \
242  05/20/1882  08/26/1882     NL   
243  04/28/1873  07/23/1873    NaN   
244  09/11/1887  09/11/1887     AA   
245  09/22/1890  09/22/1890     NL   
246  09/02/1884  09/15/1884     UA   
247  08/20/2017  08/20/2017     NL   
248  09/11/1898  09/17/1899     NL   
249  05/01/1880  09/29/1882     NL   
250  08/17/1887  08/17/1887     NL   
251  10/30/1874  10/30/1874    NaN   

                                                      notes  
242                                                     TRN  
243                                                     ELI  
244                                                     NY4  
245                                                     PIT  
246                                            9/15 forfeit  
247                                                     PIT  
248  BRO:9/18&10/2/1898; NY1:9/11/98, 6/4&7/16&8/13&9/17/99  
249                                                     NaN  
250                                              1 BSN game  
251                                              1 BS1 game


This seems to be a list of all baseball parks. There are IDs which seem to match with the game log, as well as names, nicknames, city and league.



In [10]:
person = pd.read_csv('person_codes.csv')
person.shape

(20494, 7)

In [11]:
person.head(10)

id         last     first player_debut mgr_debut coach_debut  \
0  aardd001      Aardsma     David   04/06/2004       NaN         NaN   
1  aaroh101        Aaron      Hank   04/13/1954       NaN         NaN   
2  aarot101        Aaron    Tommie   04/10/1962       NaN  04/06/1979   
3  aased001         Aase       Don   07/26/1977       NaN         NaN   
4  abada001         Abad      Andy   09/10/2001       NaN         NaN   
5  abadf001         Abad  Fernando   07/28/2010       NaN         NaN   
6  abadj101       Abadie      John   04/26/1875       NaN         NaN   
7  abbae101  Abbaticchio        Ed   09/04/1897       NaN         NaN   
8  abbeb101        Abbey      Bert   06/14/1892       NaN         NaN   
9  abbec101        Abbey   Charlie   08/16/1893       NaN         NaN   

    ump_debut  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
5         NaN  
6         NaN  
7         NaN  
8         NaN  
9  07/21/1897

In [12]:
person.tail(10)

id      last   first player_debut mgr_debut coach_debut ump_debut
20484  zulej001    Zuleta   Julio   04/06/2000       NaN         NaN       NaN
20485  zumaj001    Zumaya    Joel   04/03/2006       NaN         NaN       NaN
20486  zunim001    Zunino    Mike   06/12/2013       NaN         NaN       NaN
20487  zupcb001    Zupcic     Bob   09/07/1991       NaN         NaN       NaN
20488  zupof101      Zupo   Frank   07/01/1957       NaN         NaN       NaN
20489  zuvep001   Zuvella    Paul   09/04/1982       NaN  04/02/1996       NaN
20490  zuveg101  Zuverink  George   04/21/1951       NaN         NaN       NaN
20491  zwild101  Zwilling   Dutch   08/14/1910       NaN  04/15/1941       NaN
20492  zycht001      Zych    Tony   09/04/2015       NaN         NaN       NaN
20493  thoma102  Thompson     NaN          NaN       NaN         NaN       NaN


This seems to be a list of people with IDs. The IDs  in this table and the ones in game seems to match up. There are other interesting information like debut dates, for players, managers, coaches and umpires. We can see that some people might have been one or more of these roles.

It also looks like coaches and managers are two different things in baseball. After some research, managers are what would be called a 'coach' or 'head coach' in other sports, and coaches are more specialized, like base coaches. It also seems like coaches aren't recorded in the game log.


In [13]:
team = pd.read_csv("team_codes.csv")
team.shape


(150, 8)

In [14]:
team.head(10)

team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1    NaN   1872  1874  Baltimore         Canaries       BL1    1
5     BL2     AA   1882  1891  Baltimore          Orioles       BL2    1
6     BLN     NL   1892  1899  Baltimore          Orioles       BL2    2
7     BL4    NaN   1873  1873  Baltimore        Marylands       BL4    1
8     BLA     AL   1901  1902  Baltimore          Orioles       BLA    1
9     NYA     AL   1903     0   New York          Yankees       BLA    2

In [15]:

team.tail(10)

team_id league  start   end        city   nickname franch_id  seq
140     WS3    NaN   1871  1872  Washington   Olympics       WS3    1
141     WS4    NaN   1872  1872  Washington  Nationals       WS4    1
142     WS5    NaN   1873  1873  Washington  Nationals       WS5    1
143     WS6    NaN   1875  1875  Washington  Nationals       WS6    1
144     WS7     AA   1884  1884  Washington  Nationals       WS7    1
145     WS8     NL   1886  1889  Washington   Senators       WS8    1
146     WS9     AA   1891  1891  Washington   Senators       WS9    1
147     WSN     NL   1892  1899  Washington   Senators       WS9    2
148     WSU     UA   1884  1884  Washington  Nationals       WSU    1
149     MIA     NL   2012     0       Miami    Marlins       FLO    2


This seems to be a list of all teams, with team_ids which seem to match the game log. Interestingly, there is a franch_id, let's take a look at this:


From the above data dump the following can be inferred

   - **`team`** seems to contain all the teams represented in the game data captured in game.csv. 
       - **`team_id`** is the unique code identifiying each team
   - **`park`** seems to contain all the fields/parks/stadiums where the games were played. 
       - **`park_id`** is the unique code identifiying each team
   - **`person`** seems to contain all the players who were part of games whose data was captured in game.csv
       - **`person_id`** is the unique code identifiying each team
       
Finally, **`game`** contains all the baseball games data

### More information about the data 

**`game`**.csv contains data about [Major league baseball](https://en.wikipedia.org/wiki/Major_League_Baseball) games compiled by [Retrosheet](http://www.retrosheet.org/). Retrosheet compiles detailed statistics on baseball games from the 1800s through to today. The main file we will be working from **`game.csv`**

Each row of **`game.csv`** information about a single game, where it was playes, players and officials involved in the game and players positions. In baseball there are varios positions each players can occupy like, 
   1. Pitcher
   2. Catcher
   3. First baseman
   4. Second baseman
   5. Third baseman
   6. Shortstop
   7. Left fielder, also, less commonly known as left outfield
   8. Center fielder
   9. Right fielder

The 10th position isn't included, it may be a way of describing a designated hitter that does not field. I can find a retrosheet page that indicates that position 0 is used for this, but we don't have any position 0 in our data. I have chosen to make this an 'Unknown Position' so I'm not including data based on a hunch.

Having the right player at the right position is critical for various defence and offensive strategies that each team comes up with. You can read more about baseball positions [here](https://en.wikipedia.org/wiki/Baseball_positions)


#### Major Leagues


In [16]:
game["v_league"].value_counts(dropna=False)

NL     88866
AL     74713
AA      5039
FL      1243
NaN     1086
PL       532
UA       428
Name: v_league, dtype: int64

In [17]:
game["h_league"].value_counts(dropna=False)

NL     88867
AL     74712
AA      5039
FL      1243
NaN     1086
PL       532
UA       428
Name: h_league, dtype: int64

The game contains data from varios american baseball leagues. The league id is defined is in columns **`v_league`** and **`h_league`**. The following are the values in these columns

   - **`AA`** - [American Association](https://en.wikipedia.org/wiki/American_Association_%2819th_century%29) 
   - **`NL`** - [National League] (https://en.wikipedia.org/wiki/National_League)
   - **`AL`** - [American League] (https://en.wikipedia.org/wiki/American_League)
   - **`FL`** - [Federal League] (https://en.wikipedia.org/wiki/Federal_League)
   - **`PL`** - [Players league](https://en.wikipedia.org/wiki/Players%27_League)
   - **`UA`** - [Union Association] (https://en.wikipedia.org/wiki/Union_Association)
   

It looks like most of our games fall into the two current leagues, but that there are four other leagues. Let's write a quick function to get some info on the years of these leagues. We also have 1086 rows where there is no mention of which leage the game belonged to

In [18]:
from datetime import datetime
def league_info(league):
    league_games = game[game["h_league"] == league]
    try:
        earliest = datetime.strptime(str(league_games["date"].min()), "%Y%m%d")
        latest = datetime.strptime(str(league_games["date"].max()), "%Y%m%d")
        print("{} went from {} to {}".format(league,earliest.strftime("%b/%d/%Y"),
                                             latest.strftime("%b/%d/%Y")))
    except:
        pass
   

for league in game["h_league"].unique():
    league_info(league)

NL went from Apr/22/1876 to Oct/02/2016
AA went from May/02/1882 to Oct/06/1891
UA went from Apr/17/1884 to Oct/19/1884
PL went from Apr/19/1890 to Oct/04/1890
AL went from Apr/24/1901 to Oct/02/2016
FL went from Apr/13/1914 to Oct/03/1915


We have the date range for each league data. we can use this for further analysis

### Importing data into SQLite

Since we have already read the data into pandas dataframe we can export it to SQLite using pandas built in [DataFrame.to_sql() method] (https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) Pandas will create the table if the table does not exist. It will also infer the column data type by looking at column values. Although sometimes this will lead to wrong column types we will use it now as we will be moving data to normalized tables later

In [19]:
import sqlite3
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

db = "mlb.db"

def run_query(sql_query):
    with sqlite3.connect(db) as conn:
        return pd.read_sql(sql_query, conn)
    
def run_command(sql_command):
    with sqlite3.connect(db) as conn:
        conn.execute('PRAGMA foreign_keys = ON;')
        conn.isolation_level = None
        return conn.execute(sql_command)
    
def show_tables():
    q = '''
    SELECT
        name,
        type
    FROM sqlite_master
    WHERE type IN ("table","view");
    '''
    return run_query(q)

def delete_all_tables():
    tables = ['game','team', 'park','person',
              'person_codes','game_log', 'team_codes', 'park_codes']
    for table in tables:
        q2 = "DROP TABLE IF EXISTS {};".format(table)
        run_command(q2)



In [20]:
delete_all_tables()

In [21]:

tables = {
    "game_log": game,
    "person_codes": person,
    "team_codes": team,
    "park_codes": park
}

with sqlite3.connect(db) as conn:    
    for name, data in tables.items():
        conn.execute("DROP TABLE IF EXISTS {};".format(name))
        data.to_sql(name,conn,index=False)
        



All the tables have been created. Lets check if everything looks fine or not


In [22]:
show_tables()

name   type
0      game_log  table
1  person_codes  table
2    team_codes  table
3    park_codes  table

In [23]:
query = """
SELECT * FROM game_log LIMIT 5;
"""
temp = run_query(query)
temp

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1     None              1    FW1   
1  18710505               0         Fri    BS1     None              1    WS3   
2  18710506               0         Sat    CL1     None              2    RC1   
3  18710508               0         Mon    CL1     None              3    CH1   
4  18710509               0         Tue    BS1     None              2    TRO   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0     None              1        0        2         54.0         D       None   
1     None              1       20       18         54.0         D       None   
2     None              1       12        4         54.0         D       None   
3     None              1       12       14         54.0         D       None   
4     None              1        9        5         54.0         D       None   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0     None    None   FOR01       200.0           120.0    000000000   
1     None    None   WAS01      5000.0           145.0    107000435   
2     None    None   RCK01      1000.0           140.0    610020003   
3     None    None   CHI01      5000.0           150.0    101403111   
4     None    None   TRO01      3250.0           145.0    000002232   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   
3    077000000       46.0    15.0        2.0        1.0         2.0   10.0   
4    101003000       46.0    17.0        4.0        1.0         0.0    6.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   
3               0.0                0.0             0.0      0.0   
4               0.0                0.0             0.0      2.0   

  v_intentional_walks  v_strikeouts  v_stolen_bases v_caught_stealing  \
0                None           6.0             1.0              None   
1                None           5.0             3.0              None   
2                None           1.0             0.0              None   
3                None           1.0             0.0              None   
4                None           0.0             1.0              None   

   v_grounded_into_double v_first_catcher_interference  v_left_on_base  \
0                    -1.0                         None             4.0   
1                    -1.0                         None            12.0   
2                    -1.0                         None            10.0   
3                    -1.0                         None             7.0   
4                    -1.0                         None            12.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   
3              1.0                       6.0                 6.0   
4              1.0                       2.0                 2.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   
3             0.0      0.0       27.0       15.0      11.0             6.0   
4             0.0      0.0       27.0       12.0   

### Primary key for game dataset

The game log file does not have a single column that can be used as a primary key to uniquely identify each game. Exploring the Retrosheet site, we can find this [data dictionary](http://www.retrosheet.org/eventfile.htm) for their event files, which list every event within each game. This includes the following description:

**id**: *Each game begins with a twelve character ID record which identifies the date, home team, and number of the game. For example, ATL198304080 should be read as follows. The first three characters identify the home team (the Braves). The next four are the year (1983). The next two are the month (April) using the standard numeric notation, 04, followed by the day (08). The last digit indicates if this is a single game (0), first game (1) or second game (2) if more than one game is played during a day, usually a double header The id record starts the description of a game thus ending the description of the preceding game in the file.*

This description makes us belive that this essentially makes a custom key using the three columns. After we import the data, we'll construct this column to use as a primary key in our final database.



Now that we have all the data in the SQLite database, lets create a column for game table which will hold unique values. This column will later become primary key. Once the column has been added we will update the data in the column by concatinating date, h_name and number_of_game in the following format

**`date+h_name+number_ofgame`**




In [24]:
add_column = """
ALTER TABLE game_log
ADD COLUMN game_id TEXT
"""

try:
    run_command(add_column)
except:
    pass # we already have the column. So ignore the error

# update the new column with the format {date+h_name+number_ofgame}

update_id_query = """
UPDATE game_log
SET game_id = date || h_name || number_of_game
WHERE game_id IS NULL
"""

run_command(update_id_query)

verify_query = """
SELECT game_id, date, h_name, number_of_game FROM game_log LIMIT 10
"""
run_query(verify_query)


game_id      date h_name  number_of_game
0  18710504FW10  18710504    FW1               0
1  18710505WS30  18710505    WS3               0
2  18710506RC10  18710506    RC1               0
3  18710508CH10  18710508    CH1               0
4  18710509TRO0  18710509    TRO               0
5  18710511CL10  18710511    CL1               0
6  18710513CL10  18710513    CL1               0
7  18710513FW10  18710513    FW1               0
8  18710515FW10  18710515    FW1               0
9  18710516BS10  18710516    BS1               0

We have the column created. Our database is now ready for normalization

## Exploring Normalization opportunities

When we explore data we find the following

#### Repetition in columns

Let's look at the following segment of data. We have three columns that relate to one player, followed by three columns that relate to another player. We could restructure our data to remove this repetition

| v_player_1_id | v_player_1_name   | v_player_1_def_pos | v_player_2_id | v_player_2_name  | v_player_2_def_pos |
|---------------|-------------------|--------------------|---------------|------------------|--------------------|
| villj001      | Jonathan Villar   | 5.0                | genns001      | Scooter Gennett  | 4.0                |
| granc001      | Curtis Granderson | 8.0                | cabra002      | Asdrubal Cabrera | 6.0                |
| kendh001      | Howie Kendrick    | 7.0                | turnj001      | Justin Turner    | 5.0                |
| jasoj001      | John Jaso         | 3.0                | polag001      | Gregory Polanco  | 9.0                |
| gordd002      | Dee Gordon        | 4.0                | telit001      | Tomas Telis      | 2.0                |


#### Non-primary key columns should be attributes of the primary key
The primary key of our game log is our `game_id`, but the players name are not attributes of a game, but of the player id. If the only data we had was the game log, we would remove this column and create a new table that had the names of each player. As it happens, our `person` table already has a list of our player IDs and names, so we can remove these without the need for creating a new table first.


#### Redundant Data

Lastly, we want to eliminate any redundant data - that is, columns where the data is available elsewhere. A good example of this can be found in our `park` table, which will form the basis of our eventual `park` table. Let's look at the first few rows (we won't display the notes column as is not relevant to our discussion):


| park_id | name                          | aka                                      | city      | state | start      | end        | league |
|---------|-------------------------------|------------------------------------------|-----------|-------|------------|------------|--------|
| ALB01   | Riverside Park                | NaN                                      | Albany    | NY    | 09/11/1880 | 05/30/1882 | NL     |
| ALT01   | Columbia Park                 | NaN                                      | Altoona   | PA    | 04/30/1884 | 05/31/1884 | UA     |
| ANA01   | Angel Stadium of Anaheim      | Edison Field; Anaheim Stadium            | Anaheim   | CA    | 04/19/1966 | NaN        | AL     |
| ARL01   | Arlington Stadium             | NaN                                      | Arlington | TX    | 04/21/1972 | 10/03/1993 | AL     |
| ARL02   | Rangers Ballpark in Arlington | The Ballpark in Arlington; Ameriquest Fl | Arlington | TX    | 04/11/1994 | NaN        | AL     |





the proposed database schema is as follows


### Recommendations for normalization


The following are opportunities for normalization of our data:

  - In `person`, all the debut dates will be able to be reproduced using game log data.
  - In `team`, the start, end and sequence columns will be able to be reproduced using game log data.
  - In `park`, the start and end years will be able to be reproduced using game log data. While technically the state is an attribute of the city, we might not want to have a an incomplete city/state table so we will leave this in.
  - There are lots of places in `game log` where we have a player ID followed by the players name. We will be able to remove this and use the name data in person
  - In `game`, all offensive and defensive stats are repeated for the home team and the visiting team. We could break these out and have a table that lists each game twice, one for each team, and cut out this column repetition.
  - Similarly, in `game`, we have a listing for 9 players on each team with their positions - we can remove these and have one table that tracks player appearances and their positions.
  - We can do a similar thing with the umpires from `game`, instead of listing all four positions as columns, we can put the umpires either in their own table or make one table for players, umpires and managers.
  - We have several awards in `game` like winning pitcher and losing pitcher. We can either break these out into their own table, have a table for awards, or combine the awards in with general appearances like the players and umpires.
  
  
### Finalized database schema

<img src="https://github.com/renuka28/datascience/blob/master/schemas/baseball_schema.png?raw=true" alt="baseball games database schema" title="baseball games database schema" />



### Create tables without foreign keys



In [25]:

c1 = """
CREATE TABLE IF NOT EXISTS person (
    person_id TEXT PRIMARY KEY,
    first_name TEXT,
    last_name TEXT
);
"""

c2 = """
INSERT OR IGNORE INTO person
SELECT
    id,
    first,
    last
FROM person_codes;
"""

q = """
SELECT * FROM person
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

person_id first_name last_name
0  aardd001      David   Aardsma
1  aaroh101       Hank     Aaron
2  aarot101     Tommie     Aaron
3  aased001        Don      Aase
4  abada001       Andy      Abad

In [26]:
c1 = """
CREATE TABLE IF NOT EXISTS park (
    park_id TEXT PRIMARY KEY,
    name TEXT,
    nickname TEXT,
    city TEXT,
    state TEXT,
    notes TEXT
);
"""

c2 = """
INSERT OR IGNORE INTO park
SELECT
    park_id,
    name,
    aka,
    city,
    state,
    notes
FROM park_codes;
"""

q = """
SELECT * FROM park
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                   nickname       city state  \
0                                      None     Albany    NY   
1                                      None    Altoona    PA   
2             Edison Field; Anaheim Stadium    Anaheim    CA   
3                                      None  Arlington    TX   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX   

                                            notes  
0  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                            None  
2                                            None  
3                                            None  
4                                            None

In [27]:

c1 = """
CREATE TABLE IF NOT EXISTS league (
    league_id TEXT PRIMARY KEY,
    name TEXT
);
"""

c2 = """
INSERT OR IGNORE INTO league
VALUES
    ("NL", "National League"),
    ("AL", "American League"),
    ("AA", "American Association"),
    ("FL", "Federal League"),
    ("PL", "Players League"),
    ("UA", "Union Association")
;
"""

q = """
SELECT * FROM league
"""

run_command(c1)
run_command(c2)
run_query(q)

league_id                  name
0        NL       National League
1        AL       American League
2        AA  American Association
3        FL        Federal League
4        PL        Players League
5        UA     Union Association

In [28]:

c1 = "DROP TABLE IF EXISTS appearance_type;"

run_command(c1)

c2 = """
CREATE TABLE appearance_type (
    appearance_type_id TEXT PRIMARY KEY,
    name TEXT,
    category TEXT
);
"""
run_command(c2)

appearance_type = pd.read_csv('appearance_type.csv')

with sqlite3.connect('mlb.db') as conn:
    appearance_type.to_sql('appearance_type',
                           conn,
                           index=False,
                           if_exists='append')

q = """
SELECT * FROM appearance_type;
"""

run_query(q)

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

## Adding The Team and Game Tables

Lets add Team and Game tables now. These tables will hold tem informatoin and each individual game information

In [29]:

c1 = """
CREATE TABLE IF NOT EXISTS team (
    team_id TEXT PRIMARY KEY,
    league_id TEXT,
    city TEXT,
    nickname TEXT,
    franch_id TEXT,
    FOREIGN KEY (league_id) REFERENCES league(league_id)
);
"""

c2 = """
INSERT OR IGNORE INTO team
SELECT
    team_id,
    league,
    city,
    nickname,
    franch_id
FROM team_codes;
"""

q = """
SELECT * FROM team
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

team_id league_id       city         nickname franch_id
0     ALT        UA    Altoona  Mountain Cities       ALT
1     ARI        NL    Arizona     Diamondbacks       ARI
2     BFN        NL    Buffalo           Bisons       BFN
3     BFP        PL    Buffalo           Bisons       BFP
4     BL1      None  Baltimore         Canaries       BL1

In [30]:
c1 = """
CREATE TABLE IF NOT EXISTS game (
    game_id TEXT PRIMARY KEY,
    date TEXT,
    number_of_game INTEGER,
    park_id TEXT,
    length_outs INTEGER,
    day BOOLEAN,
    completion TEXT,
    forefeit TEXT,
    protest TEXT,
    attendance INTEGER,
    legnth_minutes INTEGER,
    additional_info TEXT,
    acquisition_info TEXT,
    FOREIGN KEY (park_id) REFERENCES park(park_id)
);
"""

c2 = """
INSERT OR IGNORE INTO game
SELECT
    game_id,
    date,
    number_of_game,
    park_id,
    length_outs,
    CASE
        WHEN day_night = "D" THEN 1
        WHEN day_night = "N" THEN 0
        ELSE NULL
        END
        AS day,
    completion,
    forefeit,
    protest,
    attendance,
    length_minutes,
    additional_info,
    acquisition_info
FROM game_log;
"""

q = """
SELECT * FROM game
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

game_id      date  number_of_game park_id  length_outs  day  \
0  18710504FW10  18710504               0   FOR01           54    1   
1  18710505WS30  18710505               0   WAS01           54    1   
2  18710506RC10  18710506               0   RCK01           54    1   
3  18710508CH10  18710508               0   CHI01           54    1   
4  18710509TRO0  18710509               0   TRO01           54    1   

  completion forefeit protest  attendance  legnth_minutes additional_info  \
0       None     None    None         200             120            None   
1       None     None    None        5000             145            HTBF   
2       None     None    None        1000             140            None   
3       None     None    None        5000             150            None   
4       None     None    None        3250             145            HTBF   

  acquisition_info  
0                Y  
1                Y  
2                Y  
3                Y  
4                Y

## Adding the Team Appearance Table¶

Lets add team_appearance table which will hold inforamtion of each appearance instance of a given team. Schema is as follows


<img src="https://github.com/renuka28/datascience/blob/master/schemas/team_appearance.png?raw=true" alt="baseball games database schema - team_appearance table" title="baseball games database schema - team_appearance table" />




In [31]:
c1 = """
CREATE TABLE IF NOT EXISTS team_appearance (
    team_id TEXT,
    game_id TEXT,
    home BOOLEAN,
    league_id TEXT,
    score INTEGER,
    line_score TEXT,
    at_bats INTEGER,
    hits INTEGER,
    doubles INTEGER,
    triples INTEGER,
    homeruns INTEGER,
    rbi INTEGER,
    sacrifice_hits INTEGER,
    sacrifice_flies INTEGER,
    hit_by_pitch INTEGER,
    walks INTEGER,
    intentional_walks INTEGER,
    strikeouts INTEGER,
    stolen_bases INTEGER,
    caught_stealing INTEGER,
    grounded_into_double INTEGER,
    first_catcher_interference INTEGER,
    left_on_base INTEGER,
    pitchers_used INTEGER,
    individual_earned_runs INTEGER,
    team_earned_runs INTEGER,
    wild_pitches INTEGER,
    balks INTEGER,
    putouts INTEGER,
    assists INTEGER,
    errors INTEGER,
    passed_balls INTEGER,
    double_plays INTEGER,
    triple_plays INTEGER,
    PRIMARY KEY (team_id, game_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id)
);
"""

run_command(c1)

c2 = """
INSERT OR IGNORE INTO team_appearance
    SELECT
        h_name,
        game_id,
        1 AS home,
        h_league,
        h_score,
        h_line_score,
        h_at_bats,
        h_hits,
        h_doubles,
        h_triples,
        h_homeruns,
        h_rbi,
        h_sacrifice_hits,
        h_sacrifice_flies,
        h_hit_by_pitch,
        h_walks,
        h_intentional_walks,
        h_strikeouts,
        h_stolen_bases,
        h_caught_stealing,
        h_grounded_into_double,
        h_first_catcher_interference,
        h_left_on_base,
        h_pitchers_used,
        h_individual_earned_runs,
        h_team_earned_runs,
        h_wild_pitches,
        h_balks,
        h_putouts,
        h_assists,
        h_errors,
        h_passed_balls,
        h_double_plays,
        h_triple_plays
    FROM game_log

UNION

    SELECT    
        v_name,
        game_id,
        0 AS home,
        v_league,
        v_score,
        v_line_score,
        v_at_bats,
        v_hits,
        v_doubles,
        v_triples,
        v_homeruns,
        v_rbi,
        v_sacrifice_hits,
        v_sacrifice_flies,
        v_hit_by_pitch,
        v_walks,
        v_intentional_walks,
        v_strikeouts,
        v_stolen_bases,
        v_caught_stealing,
        v_grounded_into_double,
        v_first_catcher_interference,
        v_left_on_base,
        v_pitchers_used,
        v_individual_earned_runs,
        v_team_earned_runs,
        v_wild_pitches,
        v_balks,
        v_putouts,
        v_assists,
        v_errors,
        v_passed_balls,
        v_double_plays,
        v_triple_plays
    from game_log;
"""

run_command(c2)

q = """
SELECT * FROM team_appearance
WHERE game_id = (
                 SELECT MIN(game_id) from game
                )
   OR game_id = (
                 SELECT MAX(game_id) from game
                )
ORDER By game_id, home;
"""

run_query(q)

team_id       game_id  home league_id  score line_score  at_bats  hits  \
0     CL1  18710504FW10     0      None      0  000000000       30     4   
1     FW1  18710504FW10     1      None      2  010010000       31     4   
2     MIA  20161002WAS0     0        NL      7  000230020       38    14   
3     WAS  20161002WAS0     1        NL     10  03023002x       30    10   

   doubles  triples  homeruns  rbi  sacrifice_hits  sacrifice_flies  \
0        1        0         0    0               0                0   
1        1        0         0    2               0                0   
2        1        1         2    7               1                0   
3        2        0         1   10               1                1   

   hit_by_pitch  walks  intentional_walks  strikeouts  stolen_bases  \
0             0      1                NaN           6             1   
1             0      1                NaN           0             0   
2             0      3                2.0          10             1   
3             1      8                0.0           3             2   

   caught_stealing  grounded_into_double  first_catcher_interference  \
0              NaN                    -1                         NaN   
1              NaN                    -1                         NaN   
2              1.0                     1                         0.0   
3              0.0                     1                         0.0   

   left_on_base  pitchers_used  individual_earned_runs  team_earned_runs  \
0             4              1                       1                 1   
1             3              1                       0                 0   
2             8              7                      10                10   
3             7              6                       7                 7   

   wild_pitches  balks  putouts  assists  errors  passed_balls  double_plays  \
0             0      0       27        9       0             3             0   
1             0      0       27        3       3             1             1   
2             1      0       24       11       0             0             1   
3             1      0       27       11       0             0             1   

   triple_plays  
0             0  
1             0  
2             0  
3             0

## Adding the Person Appearance Table¶

Now we can create our person_appearance table and add each team member information. Remember that our original csv file and table contains information about each team member appearance and we need to pull in all the information for each appearance. It is also possible that the same individual has take up multiple roles in a given game. So we will need series of select statements which will be unioned to product the final table which lists all person appeareances in various games and roles




<img src="https://github.com/renuka28/datascience/blob/master/schemas/peson_apperance.png?raw=true" alt="baseball games database schema - person_appearance table" title="baseball games database schema - person_appearance table" />

In [32]:
c0 = "DROP TABLE IF EXISTS person_appearance"

run_command(c0)

c1 = """
CREATE TABLE person_appearance (
    appearance_id INTEGER PRIMARY KEY,
    person_id TEXT,
    team_id TEXT,
    game_id TEXT,
    appearance_type_id,
    FOREIGN KEY (person_id) REFERENCES person(person_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (appearance_type_id) REFERENCES appearance_type(appearance_type_id)
);
"""

c2 = """
INSERT OR IGNORE INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        NULL,
        hp_umpire_id,
        "UHP"
    FROM game_log
    WHERE hp_umpire_id IS NOT NULL    

UNION

    SELECT
        game_id,
        NULL,
        [1b_umpire_id],
        "U1B"
    FROM game_log
    WHERE "1b_umpire_id" IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        [2b_umpire_id],
        "U2B"
    FROM game_log
    WHERE [2b_umpire_id] IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        [3b_umpire_id],
        "U3B"
    FROM game_log
    WHERE [3b_umpire_id] IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        lf_umpire_id,
        "ULF"
    FROM game_log
    WHERE lf_umpire_id IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        rf_umpire_id,
        "URF"
    FROM game_log
    WHERE rf_umpire_id IS NOT NULL

UNION

    SELECT
        game_id,
        v_name,
        v_manager_id,
        "MM"
    FROM game_log
    WHERE v_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        h_name,
        h_manager_id,
        "MM"
    FROM game_log
    WHERE h_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_pitcher_id,
        "AWP"
    FROM game_log
    WHERE winning_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score < v_score THEN h_name
            ELSE v_name
            END,
        losing_pitcher_id,
        "ALP"
    FROM game_log
    WHERE losing_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        saving_pitcher_id,
        "ASP"
    FROM game_log
    WHERE saving_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_rbi_batter_id,
        "AWB"
    FROM game_log
    WHERE winning_rbi_batter_id IS NOT NULL

UNION

    SELECT
        game_id,
        v_name,
        v_starting_pitcher_id,
        "PSP"
    FROM game_log
    WHERE v_starting_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        h_name,
        h_starting_pitcher_id,
        "PSP"
    FROM game_log
    WHERE h_starting_pitcher_id IS NOT NULL;
"""

template = """
INSERT INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "O{num}"
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL

UNION

    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "D" || CAST({hv}_player_{num}_def_pos AS INT)
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL;
"""

run_command(c1)
run_command(c2)

for hv in ["h","v"]:
    for num in range(1,10):
        query_vars = {
            "hv": hv,
            "num": num
        }
        run_command(template.format(**query_vars))

In [33]:

print(run_query("SELECT COUNT(DISTINCT game_id) games_game FROM game"))
print(run_query("SELECT COUNT(DISTINCT game_id) games_person_appearance FROM person_appearance"))

q = """
SELECT
    pa.*,
    at.name,
    at.category
FROM person_appearance pa
INNER JOIN appearance_type at on at.appearance_type_id = pa.appearance_type_id
WHERE PA.game_id = (
                   SELECT max(game_id)
                    FROM person_appearance
                   )
ORDER BY team_id, appearance_type_id
"""

run_query(q)

   games_game
0      171907
   games_person_appearance
0                   171907


appearance_id person_id team_id       game_id appearance_type_id  \
0         1646109  porta901    None  20161002WAS0                U1B   
1         1646108  onorb901    None  20161002WAS0                U2B   
2         1646107  kellj901    None  20161002WAS0                U3B   
3         1646110  tumpj901    None  20161002WAS0                UHP   
4         1646111  brica001     MIA  20161002WAS0                ALP   
5         6716279  koeht001     MIA  20161002WAS0                 D1   
6         4744553  telit001     MIA  20161002WAS0                 D2   
7         5589581  bourj002     MIA  20161002WAS0                 D3   
8         4462877  gordd002     MIA  20161002WAS0                 D4   
9         5026229  pradm001     MIA  20161002WAS0                 D5   
10        6434609  hecha001     MIA  20161002WAS0                 D6   
11        5871257  scrux001     MIA  20161002WAS0                 D7   
12        5307905  yelic001     MIA  20161002WAS0                 D8   
13        6152933  hoodd001     MIA  20161002WAS0                 D9   
14        1646113  mattd001     MIA  20161002WAS0                 MM   
15        4462878  gordd002     MIA  20161002WAS0                 O1   
16        4744554  telit001     MIA  20161002WAS0                 O2   
17        5026230  pradm001     MIA  20161002WAS0                 O3   
18        5307906  yelic001     MIA  20161002WAS0                 O4   
19        5589582  bourj002     MIA  20161002WAS0                 O5   
20        5871258  scrux001     MIA  20161002WAS0                 O6   
21        6152934  hoodd001     MIA  20161002WAS0                 O7   
22        6434610  hecha001     MIA  20161002WAS0                 O8   
23        6716280  koeht001     MIA  20161002WAS0                 O9   
24        1646112  koeht001     MIA  20161002WAS0                PSP   
25        1646116  melam001     WAS  20161002WAS0                ASP   
26        1646115  difow001     WAS  20161002WAS0                AWB   
27        1646117  schem001     WAS  20161002WAS0                AWP   
28        4181201  schem001     WAS  20161002WAS0                 D1   
29        3899525  lobaj001     WAS  20161002WAS0                 D2   
30        2772821  zimmr001     WAS  20161002WAS0                 D3   
31        3336173  difow001     WAS  20161002WAS0                 D4   
32        3054497  drews001     WAS  20161002WAS0                 D5   
33        3617849  espid001     WAS  20161002WAS0                 D6   
34        2209469  reveb001     WAS  20161002WAS0                 D7   
35        1927793  turnt001     WAS  20161002WAS0                 D8   
36        2491145  harpb003     WAS  20161002WAS0                 D9   
37        1646114  baked002     WAS  20161002WAS0                 MM   
38        1927794  turnt001     WAS  20161002WAS0                 O1   
39        2209470  reveb001     WAS  20161002WAS0                 O2   
40        2491146  harpb003     WAS  20161002WAS0                 O3   
41        2772822  zimmr001     WAS  20161002WAS0                 O4   
42        3054498  drews001     WAS  20161002WAS0                 O5   
43        3336174  difow001     WAS  20161002WAS0                 O6   
44        3617850  espid001     WAS  20161002WAS0                 O7   
45        3899526  lobaj001     WAS  20161002WAS0                 O8   
46        4181202  schem001     WAS  20161002WAS0                 O9   
47        1646118  schem001     WAS  20161002WAS0                PSP   

                  name category  
0           First Base   umpire  
1          Second Base   umpire  
2           Third Base   umpire  
3           Home Plate   umpire  
4       Losing Pitcher    award  
5              Pitcher  defense  
6              Catcher  defense  
7             1st Base  defense  
8             2nd Base  defense  
9             3rd Base  defense  
10           Shortstop  defense  
11          Left Field  defense  
12        Center Field  defense  

the above concludes creation of tables and migrating data from our original tables. Our tables are in a better format now, with very less data duplication. Breaking down of data also allows us to query data in very effecient manner.

Now that we have our database in a normalized format, we can go ahead and remvoe the tables that we imported and no longer need

## Removing the Original Tables


In [34]:
show_tables()

name   type
0            game_log  table
1        person_codes  table
2          team_codes  table
3          park_codes  table
4              person  table
5                park  table
6              league  table
7     appearance_type  table
8                team  table
9                game  table
10    team_appearance  table
11  person_appearance  table

In [35]:
tables = [
    "game_log",
    "park_codes",
    "team_codes",
    "person_codes"
]

for t in tables:
    c = '''
    DROP TABLE {}
    '''.format(t)
    
    run_command(c)

show_tables()

name   type
0             person  table
1               park  table
2             league  table
3    appearance_type  table
4               team  table
5               game  table
6    team_appearance  table
7  person_appearance  table